modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java

```diff
@@ -43,14 +43,20 @@ public static int indexToCompact(long[] keyRevisions, long compactionRevisionInc
         int i = binarySearch(keyRevisions, compactionRevisionInclusive);
 
         if (i < 0) {
-            if (i == -1) {
+            i = ~i;
+
+            if (i == 0) {
                 return NOT_FOUND;
             }
 
-            i = -(i + 2);
+            i--;
         }
 
-        if (i == keyRevisions.length - 1 && !isTombstone.test(keyRevisions[i])) {
+        if (!isTombstone.test(keyRevisions[i])) {
```

Comment:
I would propose to invert this condition, this will make the code more readable

Reply:
Also, if `keyRevisions[i]` is a tombstone, but the `compactionRevisionInclusive` was not found in the keyRevisions array, then we would return the index of the first element that is greater `compactionRevisionInclusive`. Doesn't it look strange to you?

Reply:
> we would return the index of the first element that is greater compactionRevisionInclusive.

I don't see how you got this conclusion, can you please elaborate?

Reply:
I forgot that you decrease `i` in this scenario, my bad